File: library/src/main/java/com/github/ushiosan23/android_networkutils/http/HttpConnector.java
package com.github.ushiosan23.android_networkutils.http;

import android.content.Context;

import com.android.volley.RequestQueue;
import com.android.volley.toolbox.Volley;
import com.github.ushiosan23.android_networkutils.exceptions.UnInitializePropertyException;
import com.github.ushiosan23.android_networkutils.functions.IRequire0;
import com.github.ushiosan23.android_networkutils.functions.IRequire1;

/**
 * Create a chanel to make http request.
 */
public final class HttpConnector {

	/* ---------------------------------------------------------
	 *
	 * Properties
	 *
	 * --------------------------------------------------------- */

	/**
	 * Volley request queue.
	 */
	private static RequestQueue queue;

	/* ---------------------------------------------------------
	 *
	 * Static methods
	 *
	 * --------------------------------------------------------- */

	/**
	 * Initialize http connector.
	 * I recommend to initialize this class in Application class.
	 *
	 * @param context Application context.
	 */
	public static void initialize(Context context) {
		queue = Volley.newRequestQueue(context);
	}

	/**
	 * Get current request queue.
	 *
	 * @return {@link RequestQueue} Current request instance.
	 */
	static RequestQueue getQueue() {
		return requireQueue(() -> queue);
	}

	/**
	 * Check if current request queue was initialized.
	 *
	 * @param action Action to launch.
	 * @param <T>    Object to return.
	 * @return {@link T} instance.
	 */
	public static <T> T requireQueue(IRequire1<T> action) {
		if (queue == null)
			throw new UnInitializePropertyException("Connector is not initialized.");

		return action.launch();
	}

	/**
	 * Check if current request queue was initialized.
	 *
	 * @param action Action to launch.
	 */
	public static void requireQueue(IRequire0 action) {
		if (queue == null)
			throw new UnInitializePropertyException("Connector is not initialized.");

		action.launch();
	}

}

File: app/src/main/java/com/github/ushiosan23/android/networkutils/http/HttpConnector.kt
package com.github.ushiosan23.android.networkutils.http

import android.content.Context
import com.android.volley.RequestQueue
import com.android.volley.toolbox.Volley

/**
 * Http connector object.
 *
 * This object is used to manage basic http methods
 */
object HttpConnector {

    /* ---------------------------------------------------------
     *
     * Internal properties
     *
     * --------------------------------------------------------- */

    /**
     * Create volley client
     */
    private var volleyClient: RequestQueue? = null

    /* ---------------------------------------------------------
     *
     * Public methods
     *
     * --------------------------------------------------------- */

    /**
     * Initialize connector
     *
     * This method is recommended to call on application creation.
     *
     * @param context Application context
     */
    fun initializeConnector(context: Context) {
        if (volleyClient != null) return
        volleyClient = Volley.newRequestQueue(context)
    }

    /* ---------------------------------------------------------
     *
     * Internal methods
     *
     * --------------------------------------------------------- */

    /**
     * Get instance client
     *
     * @return [RequestQueue] volley client
     */
    internal fun getClient(): RequestQueue = requireClientReturn {
        return@requireClientReturn volleyClient!!
    }

    /**
     * Require client and launch only if is initialized
     *
     * @param block Body invocation
     *
     * @throws UninitializedPropertyAccessException Error if client is not initialized
     */
    internal fun requireClient(block: () -> Unit) {
        if (volleyClient == null)
            errorInitializeClient()
        block.invoke()
    }

    /**
     * Require client and launch only if is initialized
     *
     * @param block Body invocation
     * @param T Type generic
     *
     * @return [T] generic type
     * @throws UninitializedPropertyAccessException Error if client is not initialized
     */
    internal fun <T> requireClientReturn(block: () -> T): T {
        if (volleyClient == null)
            errorInitializeClient()
        return block()
    }

    /**
     * Error if client is not initialized
     *
     * @throws UninitializedPropertyAccessException Error if client is not initialized
     */
    private fun errorInitializeClient() {
        throw UninitializedPropertyAccessException("${this::class.java.simpleName} has not been initialized.")
    }

}
